core/src/main/java/io/questdb/griffin/engine/join/AbstractAsOfJoinFastRecordCursor.java

```diff
@@ -140,22 +148,27 @@ protected long binarySearch(long masterTimestamp, long rowLo, long rowHi) {
                 if (lo < mid) {
                     lo = mid;
                 } else {
+                    // special case: when lo == mid, we're down to two elements
+                    // check if the higher element exceeds the master timestamp
                     slaveCursor.recordAt(slaveRecA, Rows.toRowID(slaveFrameIndex, hi));
                     if (slaveRecA.getTimestamp(slaveTimestampIndex) > masterTimestamp) {
-                        return lo;
+                        return lo; // hi's timestamp exceeds master, so return lo
                     }
-                    return hi;
+                    return hi; // Both elements are <= master, return the higher one
                 }
             } else {
                 hi = mid;
             }
         }
 
+        assert lo == hi;
```

Comment:
nit: this assertion doesn't look valuable yet it adds a branch even when assertions are disabled